File: dependencies/meme_4.10.0/website/src/au/edu/uq/imb/memesuite/servlet/Glam2.java
package au.edu.uq.imb.memesuite.servlet;

import au.edu.uq.imb.memesuite.data.Alphabet;
import au.edu.uq.imb.memesuite.data.SequenceDataSource;
import au.edu.uq.imb.memesuite.servlet.util.*;
import au.edu.uq.imb.memesuite.template.HTMLSub;
import au.edu.uq.imb.memesuite.template.HTMLTemplate;
import au.edu.uq.imb.memesuite.util.FileCoord;
import au.edu.uq.imb.memesuite.util.JsonWr;

import java.io.*;
import java.util.*;

import javax.activation.DataSource;
import javax.servlet.*;
import javax.servlet.http.*;

import static au.edu.uq.imb.memesuite.servlet.util.WebUtils.*;


public class Glam2 extends SubmitJob<Glam2.Data> {
  private HTMLTemplate tmplMain;
  private HTMLTemplate tmplVerify;
  private ComponentHeader header;
  private ComponentSequences sequences;
  private ComponentJobDetails jobDetails;
  private ComponentAdvancedOptions advancedOptions;
  private ComponentSubmitReset submitReset;
  private ComponentFooter footer;

  
  protected class Data extends SubmitJob.JobData {
    public String email;
    public String description;
    public SequenceDataSource seq;
    public int minAlignedSeqs;
    public int initialAlignedCols;
    public int minAlignedCols;
    public int maxAlignedCols;
    public double deletePseudo;
    public double noDeletePseudo;
    public double insertPseudo;
    public double noInsertPseudo;
    public int replicates;
    public int iterations;
    public boolean norc;
    public boolean shuffle;
    public boolean embed;

    @Override
    public void outputJson(JsonWr out) throws IOException {
      out.startObject();
      out.property("seq", seq);
      out.property("minAlignedSeqs", minAlignedSeqs);
      out.property("initialAlignedCols", initialAlignedCols);
      out.property("minAlignedCols", minAlignedCols);
      out.property("maxAlignedCols", maxAlignedCols);
      out.property("deletePseudo", deletePseudo);
      out.property("noDeletePseudo", noDeletePseudo);
      out.property("insertPseudo", insertPseudo);
      out.property("noInsertPseudo", noInsertPseudo);
      out.property("replicates", replicates);
      out.property("iterations", iterations);
      out.property("norc", norc);
      out.property("shuffle", shuffle);
      out.property("embed", embed);
      out.endObject();
    }

    @Override
    public String email() {
      return email;
    }

    @Override
    public String description() {
      return description;
    }

    @Override
    public String emailTemplate() {
      return tmplVerify.getSubtemplate("message").toString();
    }
    
    @Override
    public String cmd() {
      StringBuilder args = new StringBuilder();
      addArgs(args, "-alpha", argAlpha(seq.guessAlphabet()),
          "-min_seqs", minAlignedSeqs,
          "-min_cols", minAlignedCols,
          "-max_cols", maxAlignedCols,
          "-initial_cols", initialAlignedCols,
          "-runs", replicates,
          "-run_no_impr", iterations,
          "-del_pseudo", deletePseudo,
          "-no_del_pseudo", noDeletePseudo,
          "-ins_pseudo", insertPseudo,
          "-no_ins_pseudo", noInsertPseudo);
      if (!norc && seq.guessAlphabet() != Alphabet.PROTEIN) {
        addArgs(args, "-rev_comp");
      }
      if (embed) addArgs(args, "-embed");
      addArgs(args, seq.getName());
      return args.toString();
    }

    @Override
    public List<DataSource> files() {
      List<DataSource> sources = new ArrayList<DataSource>();
      if (seq != null) sources.add(seq);
      return sources;
    }

    @Override
    public void cleanUp() {
      if (seq != null) seq.getFile().delete();
    }
  }

  public Glam2() {
    super("GLAM2", "GLAM2");
  }

  @Override
  public void init() throws ServletException {
    super.init();
    // load the template
    this.tmplMain = cache.loadAndCache("/WEB-INF/templates/glam2.tmpl");
    this.tmplVerify = cache.loadAndCache("/WEB-INF/templates/glam2_verify.tmpl");
    header = new ComponentHeader(cache, msp.getVersion(), tmplMain.getSubtemplate("header"));
    sequences = new ComponentSequences(context, tmplMain.getSubtemplate("sequences"));
    jobDetails = new ComponentJobDetails(cache);
    advancedOptions = new ComponentAdvancedOptions(cache);
    submitReset = new ComponentSubmitReset(cache);
    footer = new ComponentFooter(cache, msp);
  }

  @Override
  public String title() {
    return tmplVerify.getSubtemplate("title").toString();
  }

  @Override
  public String subtitle() {
    return tmplVerify.getSubtemplate("subtitle").toString();
  }

  @Override
  public String logoPath() {
    return tmplVerify.getSubtemplate("logo").toString();
  }

  @Override
  public String logoAltText() {
    return tmplVerify.getSubtemplate("alt").toString();
  }

  @Override
  protected void displayForm(HttpServletRequest request, HttpServletResponse response) throws IOException {
    HTMLSub main = this.tmplMain.toSub();
    main.set("help", new HTMLSub[]{header.getHelp(), sequences.getHelp(),
        jobDetails.getHelp(), advancedOptions.getHelp(), submitReset.getHelp(),
        footer.getHelp()});
    main.set("header", header.getComponent());
    main.set("sequences", sequences.getComponent(
        request.getParameter("sequences_embed"),
        request.getParameter("sequences_name")));
    main.set("job_details", jobDetails.getComponent());
    main.set("advanced_options", advancedOptions.getComponent());
    main.set("submit_reset", submitReset.getComponent());
    main.set("footer", footer.getComponent());
    // set values from a GLAM2 rerun action
    main.set("min_seqs", WebUtils.paramOptInteger(request, "min_seqs", 2, null, 2));
    main.set("initial_cols", WebUtils.paramOptInteger(request, "initial_cols", 2, 300, 20));
    main.set("min_cols", WebUtils.paramOptInteger(request, "min_cols", 2, 300, 2));
    main.set("max_cols", WebUtils.paramOptInteger(request, "max_cols", 2, 300, 50));
    main.set("pseudo_del", WebUtils.paramOptNumber(request, "pseudo_del", 0.0, null, 0.1));
    main.set("pseudo_nodel", WebUtils.paramOptNumber(request, "pseudo_nodel", 0.0, null, 2.0));
    main.set("pseudo_ins", WebUtils.paramOptNumber(request, "pseudo_ins", 0.0, null, 0.02));
    main.set("pseudo_noins", WebUtils.paramOptNumber(request, "pseudo_noins", 0.0, null, 1.0));
    main.set("replicates", WebUtils.paramOptInteger(request, "replicates", 1, 100, 10));
    main.set("max_iter", WebUtils.paramOptInteger(request, "max_iter", 1, 1000000, 2000));
    if (WebUtils.paramBool(request, "norc")) main.set("norc_checked", "checked");
    // output
    response.setContentType("text/html");
    main.output(response.getWriter());
  }

  @Override
  protected Data checkParameters(FeedbackHandler feedback,
      HttpServletRequest request) throws IOException, ServletException {
    // setup default file names
    FileCoord namer = new FileCoord();
    FileCoord.Name posSeqName = namer.createName("sequences.fa");
    namer.createName("description");
    namer.createName("address_file");
    // create the job data
    Data data =  new Data();
    boolean error = true;
    try {
      // get the email
      data.email = jobDetails.getEmail(request, feedback);
      // get the description
      data.description = jobDetails.getDescription(request);
      // get the positive sequences
      data.seq = (SequenceDataSource)sequences.getSequences(posSeqName, request, feedback);
      // get the minimum aligned sequences
      data.minAlignedSeqs = paramInteger(feedback, "minimum aligned sequences",
          request, "min_seqs", 2, null, 2);
      // get the initial number of aligned columns
      data.initialAlignedCols = paramInteger(feedback, "initial aligned columns",
          request, "initial_cols", 2, 300, 0);
      // get the minimum number of aligned columns
      data.minAlignedCols = paramInteger(feedback, "minimum aligned columns",
          request, "min_cols", 2, 300, 0);
      // get the maximum number of aligned columns
      data.maxAlignedCols = paramInteger(feedback, "maximum aligned columns",
          request, "max_cols", 2, 300, 0);
      // check min is smaller than max
      if (data.minAlignedCols != 0 && data.maxAlignedCols != 0 && data.minAlignedCols > data.maxAlignedCols) {
        feedback.whine("The minimum number of aligned columns must be &le; " + 
            "the maximum number of aligned columns.");
      }
      // check initial is in range
      if (data.initialAlignedCols != 0) {
        if (data.minAlignedCols != 0 && data.minAlignedCols > data.initialAlignedCols) {
          feedback.whine("The initial number of aligned columns must be &ge; " +
              "the minimum number of aligned columns.");
        }
        if (data.maxAlignedCols != 0 && data.maxAlignedCols < data.initialAlignedCols) {
          feedback.whine("The initial number of aligned columns must be &le; " +
              "the maximum number of aligned columns.");
        }
      }
      // get the delete pseudocount
      data.deletePseudo = paramNumber(feedback, "pseudocount for delete",
          request, "pseudo_del", 0.0, null, 0.1);
      // get the no-delete pseudocount
      data.noDeletePseudo = paramNumber(feedback, "pseudocount for no-delete",
          request, "pseudo_nodel", 0.0, null, 2.0);
      // get the insert pseudocount
      data.insertPseudo = paramNumber(feedback, "pseudocount for insert",
          request, "pseudo_ins", 0.0, null, 0.02);
      // get the no-insert pseudocount
      data.noInsertPseudo = paramNumber(feedback, "pseudocount for no-insert",
          request, "pseudo_noins", 0.0, null, 1.0);
      // get the number of replicates
      data.replicates = paramInteger(feedback, "number of alignment replicates",
          request, "replicates", 1, 100, 10);
      // get the number of iterations without improvement
      data.iterations = paramInteger(feedback, "number of iterations without improvement",
          request, "max_iter", 1, 1000000, 2000);
      // get the flags
      data.norc = paramBool(request, "norc");
      data.shuffle = paramBool(request, "shuffle");
      data.embed = paramBool(request, "embed");
      error = false;
    } finally {
      if (error) data.cleanUp();
    }
    return data;
  }

  private static String argAlpha(Alphabet alph) {
    if (alph == null) throw new IllegalArgumentException("Alphabet should not be null");
    switch (alph) {
      case DNA:
      case RNA:
        return "DNA";
      case PROTEIN:
        return "PROTEIN";
      default:
        throw new Error("Unknown alphabet");
    }
  }
}

File: dependencies/meme_4.10.0/website/js/glam2.js

var sequences = null;

function register_component(id, element, controler) {
  "use strict";
  if (id == "sequences") {
    sequences = controler;
  }
}

function check() {
  "use strict";
  var alphabet, initial_cols, min_cols, max_cols;
  if (sequences != null) {
    if (!sequences.check()) return false;
    alphabet = sequences.get_alphabet();
  } else {
    alphabet = AlphType.DNA | AlphType.RNA | AlphType.PROTEIN;
  }
  if (!check_job_details()) return false;
  if (!check_int_value("minimum aligned sequences", "min_seqs", 2, null, 2)) return false;
  if (!check_int_value("initial aligned columns", "initial_cols", 2, 300, 20)) return false;
  if (!check_int_range("minimum aligned columns", "min_cols", 2,
        "maximum aligned columns", "max_cols", 50, 2, 300)) return false;
  initial_cols = +($('initial_cols').value);
  min_cols = +($('min_cols').value);
  max_cols = +($('max_cols').value);
  if (initial_cols < min_cols || initial_cols > max_cols) {
    alert('Please input a whole number for the initial aligned columns which ' +
        ' is between or equal to the minimum (' + min_cols + ') and maximum (' +
          max_cols + ') aligned columns.\n');
    return false;
  }
  if (!check_num_value("deletion pseudocounts", "pseudo_del", 0, null, 0.1)) return false;
  if (!check_num_value("no-deletion pseudocounts", "pseudo_nodel", 0, null, 2)) return false;
  if (!check_num_value("insertion pseudocounts", "pseudo_ins", 0, null, 0.02)) return false;
  if (!check_num_value("no-insertion pseudocounts", "pseudo_noins", 0, null, 1)) return false;
  if (!check_int_value("number of replicates", "replicates", 1, 100, 10)) return false;
  if (!check_int_value("number of iterations without improvement", "max_iter", 1, 1000000, 2000)) return false;
  return true;
}

function options_changed() {
  "use strict";
  var min_seqs = +($('min_seqs').value);
  var initial_cols = +($('initial_cols').value);
  var min_cols = +($('min_cols').value);
  var max_cols = +($('max_cols').value);
  var pseudo_del = +($('pseudo_del').value);
  var pseudo_nodel = +($('pseudo_nodel').value);
  var pseudo_ins = +($('pseudo_ins').value);
  var pseudo_noins = +($('pseudo_noins').value);
  var replicates = +($('replicates').value);
  var max_iter = +($('max_iter').value);
  if (typeof min_seqs !== "number" || isNaN(min_seqs) || min_seqs != 2) return true;
  if (typeof initial_cols !== "number" || isNaN(initial_cols) || initial_cols != 20) return true;
  if (typeof min_cols !== "number" || isNaN(min_cols) || min_cols != 2) return true;
  if (typeof max_cols !== "number" || isNaN(max_cols) || max_cols != 50) return true;
  if (typeof pseudo_del !== "number" || isNaN(pseudo_del) || pseudo_del != 0.1) return true;
  if (typeof pseudo_nodel !== "number" || isNaN(pseudo_nodel) || pseudo_nodel != 2) return true;
  if (typeof pseudo_ins !== "number" || isNaN(pseudo_ins) || pseudo_ins != 0.02) return true;
  if (typeof pseudo_noins !== "number" || isNaN(pseudo_noins) || pseudo_noins != 1) return true;
  if (typeof replicates !== "number" || isNaN(replicates) || replicates != 10) return true;
  if (typeof max_iter !== "number" || isNaN(max_iter) || max_iter != 2000) return true;
  if ($("norc").checked) return true;
  if ($("shuffle").checked) return true;
  if (!$("embed").checked) return true;
  return false;
}

function options_reset(evt) {
  "use strict";
  $("min_seqs").value = 2;
  $("initial_cols").value = 20;
  $("min_cols").value = 2;
  $("max_cols").value = 50;
  $("pseudo_del").value = 0.1;
  $("pseudo_nodel").value = 2.0;
  $("pseudo_ins").value = 0.02;
  $("pseudo_noins").value = 1.0;
  $("replicates").value = 10;
  $("max_iter").value = 2000;
  $("norc").checked = false;
  $("shuffle").checked = false;
  $("embed").checked = true;
}

function fix_reset() {
}

function on_form_submit(evt) {
  if (!check()) {
    evt.preventDefault();
  }
}

function on_form_reset(evt) {
  window.setTimeout(function(evt) {
    fix_reset();
  }, 50);
}

function on_load() {
  // add listener to the form to check the fields before submit
  $("glam2_form").addEventListener("submit", on_form_submit, false);
  $("glam2_form").addEventListener("reset", on_form_reset, false);
}

// add a load
(function() {
  "use strict";
  window.addEventListener("load", function load(evt) {
    "use strict";
    window.removeEventListener("load", load, false);
    on_load();
  }, false);
})();
